modules/openxr/openxr_api.cpp

```diff
@@ -699,16 +692,35 @@ bool OpenXRAPI::setup_spaces() {
 
 	// create play space
 	{
-		if (!is_reference_space_supported(reference_space)) {
-			print_line("OpenXR: reference space ", OpenXRUtil::get_reference_space_name(reference_space), " is not supported.");
-			return false;
+		XrPosef pose_in_reference_space = identityPose;
+
+		if (is_reference_space_supported(requested_reference_space)) {
+			reference_space = requested_reference_space;
+		} else if (requested_reference_space == XR_REFERENCE_SPACE_TYPE_LOCAL_FLOOR_EXT) {
+			if (is_reference_space_supported(XR_REFERENCE_SPACE_TYPE_STAGE)) {
+				print_verbose("OpenXR: LOCAL_FLOOR space isn't supported, emulating using STAGE and LOCAL spaces.");
+				// We'll emulate LOCAL_FLOOR using the floor height from the STAGE space, but
+				// we can't initialize that here, because we need to do it after xrWaitFrame().
+				should_reset_emulated_floor_height = true;
+			} else {
+				print_verbose("OpenXR: LOCAL_FLOOR space isn't supported, emulating using LOCAL space and an estimated floor height.");
+			}
+
+			// Use an educated guess for the floor height (which will be replaced later, if
+			// the STAGE space is supported).
+			reference_space = XR_REFERENCE_SPACE_TYPE_LOCAL;
+			pose_in_reference_space.position.y = -1.5;
```

Comment:
Should we allow the user to customize this value via a project setting? Most VR games that support seated configuration allow the user to customize the height.

Reply:
That could makes sense.

Although, for a purely seated experience, you'd probably configure your game to use "Local" (rather than "Local Floor") and move the `XROrigin` up to the desired height. If you've set your reference space to "Local Floor" (unless you simply don't know the setting exists :-)) I think it's more likely to be a standing experience.

Reply:
Actually, the more I think about this, I don't think I like the idea of guessing `-1.5` or any value here.

I think it'd be better to give the developer a way to figure out that they've fallen back on `LOCAL` without a true floor height (ie. if `STAGE` isn't supported to allow us to really emulate `LOCAL_FLOOR`), then they can adjust the height of the `XROrigin` to whatever they want, and we don't need any special configuration to bake this into the `XrSpace`. That would also make it easier to make a game that works with both OpenXR and some other interface (ex WebXR), because moving the `XROrigin` is universally applicable to all XR interfaces.

What do you think?

Reply:
I think this should indeed be left up to the developer, else people are going to expect us to emulate all unsupported spaces even when we don't have a good way of doing so.

When setting an unsupported mode it should simply fail and remain on the current mode. The developer can read back the mode to see if it was successful.

Also we need to have a nice way of exposing what features are enabled. We've got a few `is_???_supported` functions because often we switch this not based on individual extensions, but on a feature set as a whole, but we should find some nice way to make this more generic.

The reality with OpenXR is that we can't make assumptions all things are supported, and developers may have very different ways of dealing with this.

Reply:
> I think this should indeed be left up to the developer, else people are going to expect us to emulate all unsupported spaces even when we don't have a good way of doing so.
> 
> When setting an unsupported mode it should simply fail and remain on the current mode. The developer can read back the mode to see if it was successful.

I think emulating `LOCAL_FLOOR` when `STAGE` is available is reasonable, because we can do it quite well (as described in the spec). It's when both `LOCAL_FLOOR` and `STAGE` are unavailable that I think we should fallback on plain old `LOCAL` without guessing some arbitrary value for the floor height (and let the developer know so they can handle it however they see fit).

Tomorrow, I'll make the changes I have in mind for that, and you guys can let me know what you think.

Reply:
Ok, I've removed the arbitrary guess at floor height in my latest push. I'll describe the changes in more detail in a new comment below...